---
title: "Profiles"
author: "Jonas BEAL"
date: "24/01/2020"
output:
  html_document:
    number_sections: yes
---

```{r setup, include=FALSE}
models <- c("BRAF_Model", "Flobak")

knitr::opts_chunk$set(echo = TRUE)

#Paths
original_path <- normalizePath(getwd())
base_folder <- dirname(dirname(original_path))
knitr::opts_knit$set(root.dir = base_folder)
```

## Objectives

Generate patients profiles from cell Lines data for further logical modelling. Datafiles have been downloaded from CellModelPassports portal (omics profiles and CRISPR/Cas9 screening) and GDSC database (drug screening)

```{r packages_and_utilities, echo=FALSE, message=FALSE}

if (!require("pacman")) install.packages("pacman")
list.of.packages <- c("eulerr","networkD3","knitr","gridExtra","mclust","diptest","moments","magrittr","tidyverse", "readxl", "cowplot", "dplyr", "tictoc", "ggpubr")
pacman::p_load(list.of.packages, character.only = TRUE)

tbl_to_df <- function(x){
  x <- x %>% as.data.frame %>%  remove_rownames %>% column_to_rownames(var="model_name")
}
rename <- dplyr::rename
select <- dplyr::select
filter <- dplyr::filter

```

# Independent omics profiles

## Cell Model Passports dataset recap

More than 1000 cell lines with several kinds of omics data: exome-sequencing, Copy Number Alterations (CNA), RNA and clinical annotations. Here we focus on melanoma and colorectal cancer.

```{r CL_import, echo=FALSE, message=FALSE, warning=FALSE}
tic()
#First import CL clinical data for further use to correlate with logical modelling results
CL_clin <- read_csv("Data/CL/Infos/model_list_20200115.csv") %>%
  select(model_id, model_name, synonyms, model_type, COSMIC_ID, BROAD_ID,
         contains("_data"), contains("tissue"), contains("cancer")) %>%
  filter(mutation_data | cnv_data | expression_data) %>%
  mutate(TCGA_label=if_else(cancer_type=="Colorectal Carcinoma", "CR_carcinoma",
                            if_else(cancer_type=="Melanoma", "SK_melanoma", cancer_type)))
cosmic_corres <- select(CL_clin,model_name,COSMIC_ID)

#Focus on two cancer types only
good_labels <- c("CR_carcinoma", "SK_melanoma")
CL_clin %<>% filter(TCGA_label %in% good_labels)
my_cell_lines <- unlist(CL_clin$model_name)
my_cell_lines_c <- unlist(CL_clin$COSMIC_ID)

#Fonction to transpose some tibbles
tibble_transpose <- function(df_input){
  df_output <- df_input %>% gather(var, value, -gene_symbol) %>%
    spread(gene_symbol, value) %>%
    rename(model_name=var) %>% 
    type_convert
}

#Import CL mutations data
CL_mut <- read_csv("Data/CL/mutations_20191101.csv") %>% 
  filter(model_name %in% my_cell_lines)

#Import and process other CL CNA data
CL_CNA <- read_csv("Data/CL/cnv_gistic_20191101.csv", skip=1) %>%
  rename(gene_symbol=X2) %>% slice(-1) %>%
  select(gene_symbol, one_of(my_cell_lines)) %>%
  tibble_transpose

#Import and process other CL RNA data
CL_RNA <- read_csv("Data/CL/rnaseq_fpkm_20191101.csv", skip=1) %>%
  rename(gene_symbol=X2) %>% slice(-c(1,2)) %>%
  select(gene_symbol, one_of(my_cell_lines)) %>%
  tibble_transpose %>%
  mutate_if(is_numeric, function(x) log2(x+1))

#Import GDSC drug screening
CL_Drugs <- read_xlsx("Data/CL/GDSC2_fitted_dose_response_15Oct19.xlsx") %>%
  .$CELL_LINE_NAME %>% unique %>%
  intersect(my_cell_lines)

#Import CRISPR/Cas9 screening
CL_CC <- read_tsv("Data/CL/03_scaledBayesianFactors.tsv") %>%
  select(-Gene) %>% colnames  %>%
  intersect(my_cell_lines)

#Genes involved in models
generate_correspondance <- function(model){
  genenames <- read.table(paste0("Models/",model,"/",model,"_namesToHugo_curated.txt"),
                          header=T,sep="\t")
  geneindex <- strsplit(as.character(genenames[,2]), split = ",") %>%
    sapply(function(l){gsub(" ","",l)})
  geneindex <- data.frame(V1 = rep(genenames[,1],
                                   sapply(geneindex, length)),
                          V2 = unlist(geneindex))
  return(geneindex)
}
node_genes_correspondance <- sapply(models, generate_correspondance, simplify = F)

generate_model_genes <- function(model){
  model_genes_all <- unique(node_genes_correspondance[[model]]$V2) %>% sub("^\\s+", "", .)
}
model_genes <- sapply(models, generate_model_genes, simplify = F)
model_genes_all <- model_genes %>% unlist %>% unname %>% unique

#Create new variables with only genes related to the models
CLmodel_mut <- CL_mut %>% filter(gene_symbol %in% model_genes_all)
CLmodel_CNA <- CL_CNA %>% select(model_name, one_of(model_genes_all))
CLmodel_RNA <- CL_RNA %>% select(model_name, one_of(model_genes_all))
```

Many patients have all omics data and even drug screening information. Only CRISPR/Cas9 is more rare:

```{r Venn1, echo=FALSE, eval=TRUE}

intersections_patients <- list(Mutations=unique(CL_mut$model_name),
                               CNA=CL_CNA$model_name,
                               RNA=CL_RNA$model_name,
                               `Drug screening`=CL_Drugs,
                               `CRISPR screening`=CL_CC)
euler(intersections_patients, shape="ellipse") %>%
  plot(quantities=list("counts"),
       fills=list(alpha = 0.5),
       main="Number of cell lines & sets of data")
```

And almost all genes related to the model have corresponding data for all omics:

```{r Venn2, echo=FALSE, eval=TRUE}
intersections_genes <- list(Mutations=intersect(unique(CL_mut$gene_symbol),
                                                unlist(model_genes_all)),
                               CNA=intersect(select(CL_CNA, -model_name) %>% colnames,
                                             unlist(model_genes_all)),
                               RNA=intersect(select(CL_RNA, -model_name) %>% colnames,
                                             unlist(model_genes_all)),
                               Model=unlist(model_genes_all)
                            )
euler(intersections_genes, shape="ellipse") %>%
  plot(quantities=list("counts"),
       fills=list(alpha = 0.5),
       main="Number of genes & sets of data")

```

# Processing pipeline

## Mutations profiles

We need to assign Boolean effects to mutations: either 0 (inactivating) or 1 (activating). A mutation can stay unassigned in absence of any evidence. Several assignment methods are proposed in the code, based on different levels of evidence. In the article we decided to focus only only the OncoKB curated database to include only high-confidence functional inference of mutations.

```{r Classical, echo=FALSE, message=FALSE, warning=FALSE}
#The small batch of code commented below is designed to prepare a file for PolyPhen batch in order to further analyze mutations
# test <- CL_mut %>% filter(grepl("p\\.[A-Z][0-9]*[A-Z]", protein_mutation)) %>%
#   select(gene_symbol, protein_mutation) %>%
#   mutate(Interm=gsub('p.','',protein_mutation),
#          A1=substr(Interm,1,1),
#          A2=substr(Interm, nchar(Interm), nchar(Interm)),
#          Pos=gsub('[A-Z]', '', Interm)) %>%
#   select(gene_symbol, Pos, A1, A2)
# write_tsv(test, 'Polyphen_input.tsv')

polyphen <- read_delim("Data/Common/pph2-full_all.txt", delim = '\t', trim_ws = T) 
colnames(polyphen)[1:4] <- c("Gene","o_pos", "o_aa1", "o_aa2")
polyphen %<>% mutate(Label=paste(Gene, paste0(o_aa1,o_pos,o_aa2), sep="_")) %>%
  mutate(Polyphen=prediction) %>% select(Label, Polyphen) %>%
  distinct

#OncoKB database for missense assignments
OncoKB <- read_delim("Data/Common/allAnnotatedVariants.txt", delim = "\t") %>%
  mutate(Label=paste(Gene, Alteration, sep="_"))

#Functionnal inference based on Tokheim et al., 2020+ method
OncoTSG <- read_delim("Data/Common/2020_pancancer.csv", delim = ";", skip = 1) %>%
  select(gene,`oncogene q-value`,`tsg q-value`)
oncogenes <- OncoTSG %>% filter(`oncogene q-value`<=0.1) %>% select(gene) %>% unlist %>% unname
tsg <- OncoTSG %>% filter(`tsg q-value`<=0.1) %>% select(gene) %>% unlist %>% unname

#The following function performs mutations (from exome-sequencing) assignemnt to 0 or 1 with several methods
mutations_assignment <- function(x){
  output <-  x %>% 
    
    #Define the label for subsequent missense mutations interpretation
    mutate(Label=paste(gene_symbol, substr(protein_mutation, 3, 100), sep="_"),
           BIN=NA_real_) %>%
    #Assign missense mutations present in OncoKB
    mutate(Method=ifelse(Label %in% OncoKB$Label, "OncoKB", NA_character_)) %>%
    mutate(BIN=ifelse(Method=="OncoKB", 
                      ifelse(grepl("Gain",OncoKB$`Mutation Effect`[match(Label, OncoKB$Label)]),
                             1,
                             ifelse(grepl("Loss",OncoKB$`Mutation Effect`[match(Label, OncoKB$Label)]),
                                    0,
                                    BIN)
                             ),
                      BIN)) %>%
    
    #Analyze inactivating mutations
    #First nonsense
    mutate(Method=if_else(grepl("p\\.[A-Z][0-9]*\\*", protein_mutation),
                          "Inactivating", Method)) %>%
    #Then stop lost
    mutate(Method=if_else(grepl("p\\.\\*[0-9]*[A-Z]", protein_mutation),
                          "Inactivating", Method)) %>%
    #And frameshift
    mutate(Method=if_else(grepl("p\\.[A-Z][0-9]*fs\\*", protein_mutation),
                          "Inactivating", Method)) %>%
    #And define the corresponding binary assignment
    mutate(BIN=if_else(Method == "Inactivating", 0, BIN)) %>%
  
  #Use Oncogene/TSG assignments
    left_join(polyphen, by='Label') %>%
    mutate(Method=ifelse(is.na(Method) & (gene_symbol %in% oncogenes | gene_symbol %in% tsg), "2020+", Method)) %>%
    mutate(BIN=ifelse(Method=="2020+" & (gene_symbol %in% oncogenes) & (grepl("damaging",Polyphen)),
                      1,
                      BIN)) %>%
    mutate(BIN=ifelse(Method=="2020+" & (gene_symbol %in% tsg) & (grepl("damaging",Polyphen)),
                      0,
                      BIN)) %>%
    
  #Write remaining categories
  mutate(Method = ifelse(is.na(Method), "Not Processed", Method)) %>%
  mutate(BIN = ifelse(is.na(BIN), "Not Assigned", BIN)) %>%
  mutate(Status=ifelse(gene_symbol %in% model_genes_all, "Model-related gene", "Out of scope"))
}

#If you want to add less stringent mutation interpretation methods (not recommanded in our case) use the commented line instead of the active one
#CL_mut <- mutations_assignment(CL_mut) 
CL_mut <- mutations_assignment(CL_mut) %>% filter(Method=="OncoKB") 
CLmodel_mut <- CL_mut %>% filter(gene_symbol %in% model_genes_all)

```

Assignement methods and their respective influence:

```{r Sankey_plots, echo=FALSE, message=FALSE, warning=FALSE}
sankey_plot <- function(x) {
  
  x <- bind_rows(select(x,Method,BIN) %>% rename(source=Method, target=BIN),select(x,BIN,Status) %>% rename(source=BIN, target=Status)) %>% table %>% as.data.frame %>% as.tibble %>% rename(value=Freq) %>%  filter(value!=0) %>% mutate(source=as.character(source), target=as.character(target)) 
  
  
  # Create nodes DF with the unique sources & targets from input
  nodes <- data.frame(unique(c(x$source,x$target)),stringsAsFactors=FALSE)

  nodes$ID <- as.numeric(rownames(nodes)) - 1 # zero-indexed IDs for sankeyNetwork
  names(nodes) <- c("name", "ID")
  nodes <- nodes %>% as.tibble %>% mutate(NodeGroup=ifelse(name %in% c("0", "1", "Not Assigned"),"Assignment",ifelse(name %in% c("Model-related gene", "Out of scope"),"Model","Method")))

  links <- inner_join(x, nodes, by = c("source"="name")) %>%
    rename(source_ID = ID) %>%
    inner_join(nodes, by = c("target"="name")) %>%
    rename(target_ID = ID) 

  # Create Sankey Plot
  sank <- sankeyNetwork(
    Links = links, Nodes = nodes, Source = "source_ID", Target = "target_ID", Value = "value", NodeID = "name", NodeGroup = "NodeGroup", units = "mutations", fontSize = 12, nodeWidth = 30
  )
  return(sank)
}

print("Sankey plots of mutation assignments depending on methods used")
sankey_plot(CL_mut)
print("Sankey plots of mutation assignments depending on methods used (restricted to model-related nodes)")
CLmodel_mut %>% sankey_plot
```

Now we can summarize patient mutations profiles after processing. In following plots we focus on model-related genes only.

```{r Mutations_profiles, echo=FALSE, message=FALSE, warning=FALSE}
#An additional processing steo is required since one gene can carry several mutation
mutation_fusion <- function(x){
  #First, we only retain distinct rows
  output_profile <- x %>% select(model_name, gene_symbol, BIN) %>% mutate(BIN=ifelse(BIN=="Not Assigned",NA,BIN)) %>% type_convert %>% distinct
  #Then, when 0&NA or 1&NA cases appear we keep only 0/1 and discard NA information
  logi <- duplicated(output_profile[,1:2]) | duplicated(output_profile[,1:2], fromLast = T)
  output_profile <- filter(output_profile, !(logi & is.na(BIN)))
  #Then, when 0&1 cases appear we keep only 0 since inactivating mutations (usually truncating) are likely to erase gain of function effects (usally achieved through missense)
  logi <- duplicated(output_profile[,1:2]) | duplicated(output_profile[,1:2], fromLast = T)
  output_profile <- filter(output_profile, !(logi & BIN==1))
  
  return(output_profile)
}

CL_mut_prof <- CL_mut %>% mutation_fusion %>% spread(gene_symbol,BIN) %>% tbl_to_df %>%
  select_if(~!all(is.na(.)))
#Restrain genes
CLmodel_mut_prof <- CL_mut_prof %>% select(one_of(model_genes_all))

# #To OncoKB
# CL_mutOKB_prof <- CL_mutOKB %>% mutation_fusion %>% spread(gene_symbol,BIN) %>%
#   tbl_to_df %>%
#   select_if(~!all(is.na(.)))
# #Restrain genes
# CLmodel_mutOKB_prof <- CL_mutOKB_prof %>% select(one_of(model_genes_all))

#Plots per genes
CLmodel_mut_prof %>% is.na %>% `!` %>% colSums %>%
  as.data.frame %>% rownames_to_column(var="Gene")  %>% rename(NumMut=".") %>%
  ggplot(aes(x=reorder(Gene, -NumMut), y=NumMut)) +
  geom_bar(stat="identity") +
  labs(x="Genes", y="Number of mutations per genes",
       title="Distribution of mutations across genes in the cohort") +
  theme_pubclean() +
  theme(axis.text.x = element_text(angle = 90))

#Plots per patients
CLmodel_mut_prof %>% is.na %>% `!` %>% rowSums %>%
  as.data.frame %>% rownames_to_column(var="Gene")  %>% rename(NumMut=".") %>%
  ggplot(aes(x=reorder(Gene, -NumMut), y=NumMut)) +
  geom_bar(stat="identity") +
  labs(x="Genes", y="Number of mutations per cell line",
       title="Distribution of mutations across cell lines in the cohort") +
  theme_pubclean() +
  theme(axis.text.x = element_text(angle = 90, size=6))

```

## CNA profiles

For CNA, we have decided to focus on stringent amplifications/deletions corresponding to +2/-2 GISTIC results. We produce the same kind of plots.

```{r CNA_profiles, echo=FALSE, message=FALSE, warning=FALSE}

CL_bin <- CL_CNA %>% select(model_name, one_of(model_genes_all)) %>% tbl_to_df %>% abs %>% equals(2)

#Plots per genes
CL_bin %>% colSums %>%
  as.data.frame %>% rownames_to_column(var="Gene")  %>% rename(NumMut=".") %>%
  ggplot(aes(x=reorder(Gene, -NumMut), y=NumMut)) + geom_bar(stat="identity") +
  labs(x="Genes", y="Number of CNA per genes",
       title="Distribution of CNA across genes in the cohort") +
  theme_pubclean() +
  theme(axis.text.x = element_text(angle = 90))

#Plots per patients
CL_bin %>% rowSums %>%
  as.data.frame %>% rownames_to_column(var="Gene")  %>% rename(NumMut=".") %>%
  ggplot(aes(x=reorder(Gene, -NumMut), y=NumMut)) + geom_bar(stat="identity") +
  labs(x="Genes", y="Number of CNA per cell line",
       title="Distribution of CNA across cell lines in the cohort") +
  theme_pubclean() +
  theme(axis.text.x = element_text(angle = 90, size=6))

#Profiles
cna_profile <- function(x){
  interm <- x %>% tbl_to_df
  interm[interm==-1 | interm==0 | interm==1] <- NA
  interm[interm==-2] <- 0
  interm[interm==2] <- 1
  return(interm)
}

CL_CNA_prof <- cna_profile(CL_CNA)
CLmodel_CNA_prof <- CL_CNA_prof %>% select(one_of(model_genes_all))

```

## RNA profiles

RNA data is intrinsically continuous and therefore require preliminary data processing. We provide here some code for both binarization and normalization of RNA data. Please note that only normalization of RNA data has been used in the article (for continuous personalization of logical models). However, the reader might be interested by binarization of continuous data and further use for discrete personalization in other applications.

### Binarization with classification tree

```{r binarization_normalization_utilities, echo=FALSE, message=FALSE, warning=FALSE}
#function to compute the Bimodality Index (BI) described in Wang et al. (2009)
BI <- function(dataset) {
  x <- dataset
    mc <- Mclust(na.omit(x), G = 2, modelNames = "E", verbose = FALSE)
    if (is.null(mc)) {
      bi <- NA
    } else {
      sigma <- sqrt(mc$parameters$variance$sigmasq)
      delta <- abs(diff(mc$parameters$mean))/sigma
      pi <- mc$parameters$pro[1]
      bi <- delta * sqrt(pi*(1-pi))
    }
  bi
}

#function to binarize the tails of the distribution, based on inter-quartile range (IQR), similar to methods described in teh outlier-sum statistic (Tibshirani and Hastie, 2007). Can be called with a reference dataset
OSclass <- function(exp_dataset, ref_dataset=exp_dataset) {
  classif <-rep(NA,length(exp_dataset))
  q25 <- quantile(ref_dataset,0.25, na.rm = T)
  q75 <- quantile(ref_dataset,0.75, na.rm = T)
  IQR <- q75 - q25 #InterQuartile Range
  classif[exp_dataset>IQR+q75] <- 1
  classif[exp_dataset<q25-IQR] <- 0
  return(classif)
}

#function to to binarize bimodal distributions based on a 2-modes gaussian mixture model (with equal variances). Can be called with a reference dataset
BIMclass <- function(exp_dataset, ref_dataset=exp_dataset) {
  mc <- Mclust(na.omit(ref_dataset), modelNames = "E", G=2, verbose = FALSE)
  classif <- rep(NA,length(exp_dataset))
  if (diff(mc$parameters$mean)>0){
    thresh_down <- max(mc$data[mc$classification==1 & mc$uncertainty <= 0.05])
    thresh_up <- min(mc$data[mc$classification==2 & mc$uncertainty <= 0.05])
    classif[exp_dataset<=thresh_down] <- 0
    classif[exp_dataset>=thresh_up] <- 1
  } else if (diff(mc$parameters$mean)<0){
    thresh_down <- max(mc$data[mc$classification==2 & mc$uncertainty <= 0.05])
    thresh_up <- min(mc$data[mc$classification==1 & mc$uncertainty <= 0.05])
    classif[exp_dataset<=thresh_down] <- 0
    classif[exp_dataset>=thresh_up] <- 1
  }
  return(classif)
}

#function for normalization of zero-inflated data
norm_fun_lin <- function(xdat, reference = xdat){
  x_proc <- (xdat-quantile(reference, 0.01, na.rm = T))/quantile(xdat-quantile(reference, 0.01, na.rm = T), 0.99, na.rm = T)
  x_proc[x_proc<0] <- 0
  x_proc[x_proc>1] <- 1
  x_proc
}

#function for normalization of unimodal data
norm_fun_sig <- function(xdat, reference = xdat){
  xdat <- xdat - median(reference, na.rm = T)
  lambda <- log(3)/mad(reference, na.rm = T)
  transformation <- function(x){
    y <- 1/(1+exp(-lambda*x))
    y
  }
  transformation(xdat) 
}

#function for normalization of unimodal data
norm_fun_bim <- function(xdat, reference = xdat) {
  not_na_xdat <- !is.na(xdat)
  not_na_ref <- !is.na(reference)
  mc <- Mclust(reference[not_na_ref], modelNames = "E", G=2, verbose = FALSE)
  pred <- predict.Mclust(mc,xdat[not_na_xdat])
  normalization <- rep(NA,length(xdat))
  if (diff(mc$parameters$mean)>0){
    normalization[not_na_xdat] <- pred$z[,2]
  } else if (diff(mc$parameters$mean)<0){
    normalization[not_na_xdat] <- pred$z[,1]
  }
  normalization
}
```

```{r RNA_profiles_original_criteria, echo=FALSE, message=FALSE, warning=FALSE, eval=TRUE}

#Here we compute all statistical tools and criteria needed to perform the classification of distributions in the following categories: discarded, zero-inflated, unimodal and bimodal
compute_criteria <- function(exp_dataset){
  exp_dataset <- exp_dataset %>% select(-model_name)
  criteria <- tibble(gene_symbol=colnames(exp_dataset), Dip=NA, BI=NA, Kurtosis=NA, DropOutRate=NA, MeanNZ=NA, DenPeak=NA, Amplitude=NA)
  
  if (dim(exp_dataset)[1]<30){
    criteria %<>% mutate(Category='Unimodal')
    return(criteria)
  }
  
  #Compute
  pb = txtProgressBar(min = 1, max = ncol(exp_dataset), initial = 1) 
  for (i in 1:ncol(exp_dataset)){
    x <- na.omit(unlist(exp_dataset[,i]))
    criteria$Amplitude[i] <- max(x)-min(x)
    
    if (criteria$Amplitude[i] !=0){
      criteria$Dip[i] <- dip.test(x)$p.value
      criteria$BI[i] <- BI(x)
      criteria$Kurtosis[i] <- kurtosis(x)-3
      criteria$DropOutRate[i] <- sum(x==0)/length(x)
      criteria$MeanNZ[i] <- sum(x)/sum(x!=0)
      den <- density(x, na.rm = T)
      criteria$DenPeak[i] <- den$x[which.max(den$y)]
    }
    
    setTxtProgressBar(pb,i)
  }
  
  threshold <- median(criteria$Amplitude)/10
  criteria <- criteria %>% 
    mutate(Category=ifelse(Amplitude<threshold | DropOutRate>0.95, "Discarded", NA)) %>%
    mutate(Category=ifelse(is.na(Category) & (BI>1.5 & Dip<0.05 & Kurtosis < 1),"Bimodal",Category)) %>%
    mutate(Category=ifelse(is.na(Category) & DenPeak<threshold, "ZeroInf", Category)) %>%
    mutate(Category=ifelse(is.na(Category), "Unimodal", Category))
  
  return(criteria)
}

#Here we compute this only for the model-related nodes. Please use the commented line below instead of the active one if you want to have a look at all genes.
#criteria_CL <- compute_criteria(CL_RNA)
criteria_CL <- select(CL_RNA, model_name, one_of(model_genes_all)) %>%
  compute_criteria
```

Now, what about the distribution of gene categories (Bimodal, Unimodal...) across the cohort?

```{r RNA_profiles_original_catDistrib, echo=FALSE, message=FALSE, warning=FALSE}
print("CL assignments for model-related nodes:")
kable(t(table(criteria_CL$Category)))

```

Here are some distributions plots randomly picked in each category in CL cohort

```{r RNA_profiles_original_catPlot, echo=FALSE, message=FALSE, warning=FALSE, eval=TRUE}
set.seed(2)
criteria_CL %>% filter(Category=="Bimodal", gene_symbol %in% model_genes_all) %>%
  select(gene_symbol) %>%
  slice(sample(nrow(.), 3)) %>% unlist %>% unname %>% select(CL_RNA,.) %>%  gather %>%
  ggplot(mapping = aes(x = value)) +
  geom_histogram(bins = 30) +
  facet_wrap(~key, scales = 'free') +
  ggtitle("3 random Bimodal genes") +
  theme_pubclean()

criteria_CL %>% filter(Category=="ZeroInf", gene_symbol %in% model_genes_all) %>%
  select(gene_symbol) %>%
  slice(sample(nrow(.), 3)) %>% unlist %>% unname %>% select(CL_RNA,.) %>%  gather %>%
  ggplot(mapping = aes(x = value)) +
  geom_histogram(bins = 30) + facet_wrap(~key, scales = 'free') +
  ggtitle("3 random Zero-Inflated genes") +
  theme_pubclean()

criteria_CL %>% filter(Category=="Unimodal", gene_symbol %in% model_genes_all) %>%
  select(gene_symbol) %>%
  slice(sample(nrow(.), 3)) %>% unlist %>% unname %>% select(CL_RNA,.) %>%  gather %>%
  ggplot(mapping = aes(x = value)) +
  geom_histogram(bins = 30) +
  facet_wrap(~key, scales = 'free') +
  ggtitle("3 random Unimodal genes") +
  theme_pubclean()

```

And depending on distribution category, we can perform binarization

```{r RNA_profiles_original_profiles, echo=FALSE, message=FALSE, warning=FALSE}
#function to apply the proper binarization method depending on the gene expression distribution category
binarize_exp <-  function(exp_dataset, ref_dataset, ref_criteria, gene, show.plot=F){
  if(!missing(gene)){
    
    gene_cat <- ref_criteria %>% filter(gene_symbol==gene) %>% select(Category) %>% unlist
    x <- unlist(select(exp_dataset,gene))
    x_ref <- unlist(select(ref_dataset,gene))
    
    if (gene_cat=="Discarded"){
        stop("Discarded gene")
      
      } else if (gene_cat=="Bimodal"){
        gene_bin <- BIMclass(x,x_ref)
        
      } else {
        gene_bin <- OSclass(x,x_ref)
      }
    names(gene_bin) <- exp_dataset$model_name
    if(show.plot==T){
      if(all(is.na(gene_bin))){
        tibble(Continuous=x) %>%
          ggplot(aes(x=Continuous)) +
          geom_histogram(bins=30) +
          ggtitle(gene) +
          theme_pubclean()
      } else {
        tibble(Continuous=x, Discrete=factor(gene_bin)) %>%
          ggplot(aes(x=Continuous, fill=Discrete)) +
          geom_histogram(bins=30) +
          ggtitle(gene) +
          theme_pubclean()
      }
    } else {
      return(gene_bin)
    }
    
  } else {
    exp_dataset <- tbl_to_df(exp_dataset) 
    ref_dataset <- tbl_to_df(ref_dataset)
    if(dim(exp_dataset)[2] != dim(ref_criteria)[1]){stop("Different number of genes")}
    logi_dis <- ref_criteria$Category=="Discarded"
    logi_OS <- ref_criteria$Category=="Unimodal" | ref_criteria$Category=="ZeroInf"
    logi_bim <- ref_criteria$Category=="Bimodal"
    exp_dataset[,logi_dis] <- lapply(exp_dataset[,logi_dis], function(x) rep(NA, length(x)))
    exp_dataset[,logi_OS] <- mapply(function(x,y) OSclass(x,y), exp_dataset[,logi_OS], ref_dataset[,logi_OS])
    exp_dataset[,logi_bim] <- mapply(function(x,y) BIMclass(x,y), exp_dataset[,logi_bim], ref_dataset[,logi_bim])
    
    return(exp_dataset)
  }
  
}

print("Bimodal example:")
binarize_exp(CL_RNA,CL_RNA, criteria_CL, "SOX10", T)
print("Zero-Inf example:")
binarize_exp(CL_RNA,CL_RNA, criteria_CL, "FGFR2", T)
print("Unimodal example:")
binarize_exp(CL_RNA,CL_RNA, criteria_CL, "NRAS", T)

#Profiles
CLmodel_RNA_prof <- binarize_exp(CLmodel_RNA, CLmodel_RNA, criteria_CL)
```

### Normalization

```{r norm_profiles, echo=FALSE, message=FALSE, warning=FALSE}

normalize_exp <-  function(exp_dataset, ref_dataset, ref_criteria, gene, show.plot=F){
  if(!missing(gene)){
    
    gene_cat <- ref_criteria %>% filter(Gene==gene) %>% select(Category) %>% unlist
    x <- unlist(select(exp_dataset,gene))
    x_ref <- unlist(select(ref_dataset,gene))
    
    if (gene_cat=="Discarded"){
        stop("Discarded gene")
      
      } else if (gene_cat=="Bimodal"){
        gene_bin <- norm_fun_bim(x,x_ref)
        
      } else if (gene_cat=="Unimodal"){
        gene_bin <- norm_fun_sig(x,x_ref)
        
      } else {
        gene_bin <- norm_fun_lin(x,x_ref)
      }
    names(gene_bin) <- exp_dataset$model_name
    
    if(show.plot==T){
      gene_bin %>% unlist %>% as.data.frame %>%
        ggplot(aes(x=.)) +
        geom_histogram(bins=30) +
        xlab(gene) +
        theme_pubclean()
    } else {
      return(gene_bin)
    }
    
  } else {
    exp_dataset <- tbl_to_df(exp_dataset) 
    ref_dataset <- tbl_to_df(ref_dataset)
    if(dim(exp_dataset)[2] != dim(ref_criteria)[1]){stop("Different number of genes")}
    logi_dis <- ref_criteria$Category=="Discarded"
    logi_uni <- ref_criteria$Category=="Unimodal"
    logi_zero <- ref_criteria$Category=="ZeroInf"
    logi_bim <- ref_criteria$Category=="Bimodal"
    exp_dataset[,logi_dis] <- lapply(exp_dataset[,logi_dis], function(x) rep(NA, length(x)))
    exp_dataset[,logi_uni] <- mapply(function(x,y) norm_fun_sig(x,y), exp_dataset[,logi_uni], ref_dataset[,logi_uni])
    exp_dataset[,logi_zero] <- mapply(function(x,y) norm_fun_lin(x,y), exp_dataset[,logi_zero], ref_dataset[,logi_zero])
    exp_dataset[,logi_bim] <- mapply(function(x,y) norm_fun_bim(x,y), exp_dataset[,logi_bim], ref_dataset[,logi_bim])
    
    return(exp_dataset)
  }
  
}

CLmodel_RNA_prof_norm <- normalize_exp(CLmodel_RNA, CLmodel_RNA, criteria_CL)

#Examples of normalization results
plot_normalization <- function(gene){
  p1 <- ggplot(CL_RNA, aes_string(gene)) +
    geom_histogram(bins=30) +
    theme_pubclean() +
    labs(title = paste0("Distribution of ", gene, " in original data"))
  p2 <- ggplot(CLmodel_RNA_prof_norm, aes_string(gene)) +
    geom_histogram(bins=30) +
    theme_pubclean() +
    labs(title = paste0("Distribution of ", gene, " in normalized data"))
  plot_grid(p1, p2, nrow = 2)
}

print("Bimodal example:")
plot_normalization("SOX10")
print("Zero-Inf example:")
plot_normalization("FGFR2")
print("Unimodal example:")
plot_normalization("NRAS")
```

# Merged profiles

## Data types relations

Before merging independant profiles into multi-omics profiles, let's have a look at relations between data types

### Mutations and CNA

In particular, is there any mutation/CNA binary inconsistency?

```{r incons_mut_CNA, echo=FALSE, message=FALSE, warning=FALSE}
common_genes <- intersect(colnames(CLmodel_CNA_prof), colnames(CLmodel_mut_prof))
common_patients <- intersect(rownames(CLmodel_CNA_prof), rownames(CLmodel_mut_prof))

incons <- which(CLmodel_mut_prof[common_patients, common_genes]==0 & CLmodel_CNA_prof[common_patients, common_genes]==1 | CLmodel_mut_prof[common_patients, common_genes]==1 & CLmodel_CNA_prof[common_patients, common_genes]==0, arr.ind = T) %>% as.data.frame %>% mutate(Patient=common_patients[row],Gene=common_genes[col], Mut=diag(as.matrix(CLmodel_mut_prof[Patient,Gene])), CNA=diag(as.matrix(CLmodel_CNA_prof[Patient,Gene]))) %>% select(Patient,Gene, Mut, CNA)

if (nrow(incons)==0){
  print("No mutation/CNA binary inconsistencies")
}else{
  kable(incons)
}
```

In case of ambiguity, pritority is given to mutations over CNA

```{r merge_mut_CNA, echo=FALSE, message=FALSE, warning=FALSE}

mut_plus_CNA <- function(mut,cna){
  merged <- mut
  merged[is.na(merged)] <- cna[is.na(merged)]
  return(merged)
}

mut_plus_CNA_big <- function(data_mut, data_CNA){
  common_patients <- intersect(rownames(data_mut), rownames(data_CNA))
  concerned_genes <- unique(c(colnames(data_mut), colnames(data_CNA)))
  
  data_merge <- matrix(nrow = length(common_patients),
                       ncol = length(concerned_genes),
                       dimnames = list(common_patients,concerned_genes)) %>%
    as.data.frame
  
  for(gene in concerned_genes){
    if (gene %in% colnames(data_mut) & gene %in% colnames(data_CNA)){
      data_merge[,gene] <- mut_plus_CNA(data_mut[common_patients,gene], data_CNA[common_patients,gene])
  } else if (gene %in% colnames(data_CNA)) {
    data_merge[,gene] <- data_CNA[common_patients,gene]
  } else {
    data_merge[,gene] <- data_mut[common_patients,gene]
  }
  }
  return(data_merge)
}

CLmodel_mutCNA_prof <- mut_plus_CNA_big(CLmodel_mut_prof, CLmodel_CNA_prof)

#Restriction with only CNAneg
CLmodel_CNAneg_prof <- CLmodel_CNA_prof
CLmodel_CNAneg_prof[CLmodel_CNA_prof==1] <- NA
CLmodel_mutCNAneg_prof <- mut_plus_CNA_big(CLmodel_mut_prof, CLmodel_CNAneg_prof)

```

### CNA and RNA

```{r incons_CNA_RNA, echo=FALSE, message=FALSE, warning=FALSE}
common_genes <- intersect(colnames(CLmodel_CNA_prof), colnames(CLmodel_RNA_prof))
common_patients <- intersect(rownames(CLmodel_CNA_prof), rownames(CLmodel_RNA_prof))

incons <- which(CLmodel_CNA_prof[common_patients, common_genes]==0 & CLmodel_RNA_prof[common_patients, common_genes]==1 | CLmodel_CNA_prof[common_patients, common_genes]==1 & CLmodel_RNA_prof[common_patients, common_genes]==0, arr.ind = T) %>% as.data.frame %>% mutate(Patient=common_patients[row],Gene=common_genes[col],  CNA=diag(as.matrix(CLmodel_CNA_prof[Patient,Gene])), RNA=diag(as.matrix(CLmodel_RNA_prof[Patient,Gene]))) %>% select(Patient,Gene,CNA, RNA)

if (nrow(incons)==0){
  print("No RNA/CNA binary inconsistencies")
}else{
  kable(incons)
}

```

## Write profiles

```{r write, echo=FALSE, message=FALSE, warning=FALSE}

inverse_mapping <- function(dataset, datatype = 'bin', geneindex_var){
  
  geneindex_var %<>% mutate(V1=as.character(V1), V2=as.character(V2))
  inv_dataset <- data.frame(row.names = rownames(dataset))
  patients <- rownames(dataset)
  nodes_list <- geneindex_var$V1 %>% unique %>% as.character
  
  for (node in nodes_list){
    mapped_genes <- geneindex_var$V2[geneindex_var$V1 == node]
    if (any(mapped_genes != "") & any(mapped_genes %in% colnames(dataset))){
      if (length(mapped_genes) == 1){
        new_column <- select(dataset, one_of(mapped_genes))
      } else if (datatype == 'norm'){
        new_column <-  data.frame(rowMeans(select(dataset, one_of(mapped_genes)), na.rm = T))
      } else if (datatype == 'bin'){
        interm <- rowMeans(select(dataset, one_of(mapped_genes)), na.rm = T)
        interm[interm!=0 & interm !=1] <- NA
        interm[is.nan(interm)] <- NA
        new_column <- data.frame(interm)
      }
      inv_dataset[patients,node] <- new_column[patients,]
    }
  }
  return(inv_dataset)
}

if (!dir.exists("Results/Profiles")){
  dir.create("Results/Profiles")
}

#Loop over models
for (model in models){
  write.csv(inverse_mapping(CLmodel_mut_prof,
                            datatype = 'bin',
                            geneindex_var = node_genes_correspondance[[model]]),
            paste0("Results/Profiles/",model,"_CL_mut.csv"))
  
  write.csv(inverse_mapping(CLmodel_CNA_prof,
                            datatype = 'bin',
                            geneindex_var = node_genes_correspondance[[model]]),
            paste0("Results/Profiles/",model,"_CL_CNA.csv"))

  write.csv(inverse_mapping(CLmodel_mutCNA_prof,
                            datatype = 'bin',
                            geneindex_var = node_genes_correspondance[[model]]),
            paste0("Results/Profiles/",model,"_CL_mutCNA.csv"))
  
  write.csv(inverse_mapping(CLmodel_mutCNAneg_prof,
                            datatype = 'bin',
                            geneindex_var = node_genes_correspondance[[model]]),
            paste0("Results/Profiles/",model,"_CL_mutCNAneg.csv"))
  
  write.csv(inverse_mapping(CLmodel_RNA_prof,
                            datatype = 'bin',
                            geneindex_var = node_genes_correspondance[[model]]),
            paste0("Results/Profiles/",model,"_CL_RNA.csv"))
  
  write.csv(inverse_mapping(CLmodel_RNA_prof_norm,
                            datatype = 'norm',
                            geneindex_var = node_genes_correspondance[[model]]),
            paste0("Results/Profiles/",model,"_CL_RNA_norm.csv"))
  
}

print("Done")
print(paste0("Computation time:"))
toc()
```

